storage/inmem/txn.go

```diff
@@ -0,0 +1,410 @@
+
+	if !txn.write {
+		return &storage.Error{
+			Code:    storage.InvalidTransactionErr,
+			Message: "data write during read transaction",
+		}
+	}
+
+	if len(path) == 0 {
+		return txn.assertRoot(op, value)
+	}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if curr.path.Equal(path) {
+			if curr.remove {
+				if op != storage.AddOp {
+					return notFoundError(path)
+				}
+			}
+			curr.stale = true
+			break
+		} else if curr.path.HasPrefix(path) {
+			curr.stale = true
+		} else if path.HasPrefix(curr.path) {
+			if curr.remove {
+				return notFoundError(path)
+			}
+			suffix := path[len(curr.path):]
+			path, remove, value, err := validate(curr.value, op, suffix, 0, value)
+			if err != nil {
+				return err
+			}
+			if len(path) == 0 {
+				curr.remove = remove
+				curr.value = value
+			} else {
+				apply(curr.value, path, remove, value)
+			}
+			return nil
+		}
+	}
+
+	path, remove, value, err := validate(txn.db.data, op, path, 0, value)
+	if err != nil {
+		return err
+	}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if curr.stale {
+			if curr == txn.log {
+				txn.log = curr.next
+			}
+			if curr.next != nil {
+				curr.next.prev = curr.prev
+			}
+			if curr.prev != nil {
+				curr.prev.next = curr.next
+			}
+		}
+	}
+
+	curr := &snapshot{
+		path:   path,
+		remove: remove,
+		value:  value,
+		next:   txn.log,
+		prev:   nil,
+	}
+
+	if txn.log != nil {
+		txn.log.prev = curr
+	}
+
+	txn.log = curr
+	return nil
+}
+
+func (txn *transaction) assertRoot(op storage.PatchOp, value interface{}) error {
+	if op == storage.RemoveOp {
+		return invalidPatchError(rootCannotBeRemovedMsg)
+	}
+	if _, ok := value.(map[string]interface{}); !ok {
+		return invalidPatchError(rootMustBeObjectMsg)
+	}
+	txn.log = &snapshot{
+		path:   storage.Path{},
+		remove: false,
+		value:  value,
+		next:   nil,
+	}
+	return nil
+}
+
+func (txn *transaction) commit() (result storage.TriggerEvent) {
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if len(curr.path) > 0 {
+			apply(txn.db.data, curr.path, curr.remove, curr.value)
+		} else {
+			txn.db.data = curr.value.(map[string]interface{})
+		}
+		result.SetDataChanged()
+	}
+	for id, bs := range txn.policies {
+		if bs == nil {
+			delete(txn.db.policies, id)
+		} else {
+			txn.db.policies[id] = bs
+		}
+		result.SetPolicyChanged()
+	}
+	return result
+}
+
+func (txn *transaction) read(path storage.Path) (interface{}, error) {
+
+	if !txn.write {
+		return ptr(txn.db.data, path)
+	}
+
+	merge := []*snapshot{}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if path.HasPrefix(curr.path) {
+			if curr.remove {
+				return nil, notFoundError(path)
+			}
+			return ptr(curr.value, path[len(curr.path):])
+		}
+		if curr.path.HasPrefix(path) {
+			merge = append(merge, curr)
+		}
+	}
+
+	data, err := ptr(txn.db.data, path)
+
+	if err != nil {
+		if !storage.IsNotFound(err) || len(merge) == 0 {
```

Comment:
I think it's correct as is; the logic is a bit confusing though.

If && was used then non-NotFound errors would only be returned if there was a matching snapshot for path.

The current logic (in pseudocode) is:

- If non-NotFound error, return err
- If NotFound error AND no matching snapshots, return err